File: src/viz/palettes.py
"""Colour palette definitions for all chart traces.

Each palette is a ``PaletteConfig`` frozen dataclass.  The ``PALETTES``
registry maps human-readable names to palette instances and is used by
the UI's palette picker dropdown.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class PaletteConfig:
    """Typed colour palette for all chart traces and UI accents.

    Attributes:
        name: Human-readable palette name.
        accent: Primary UI accent color (buttons, active states, titles,
            spinners, focus rings). Dedicated to interactive chrome —
            never used on chart traces.
        raw: Colour for raw data points and line.
        smoothed: Colour for the rolling-mean (smoothed) line.
        fit: Colour for the exponential-decay fit curve.
        derivative: Colour for negative derivative bars.
        derivative_pos: Colour for positive derivative bars.
        derivative_smooth: Colour for the smoothed derivative line.
        residual_above: Colour for plateau shading (residual > 0).
        residual_below: Colour for acceleration shading (residual < 0).
    """

    name: str
    accent: str
    raw: str
    smoothed: str
    fit: str
    derivative: str
    derivative_pos: str
    derivative_smooth: str
    residual_above: str
    residual_below: str


PALETTES: dict[str, PaletteConfig] = {
    "Classic": PaletteConfig(
        name="Classic",
        accent="#2563EB",   # blue-600
        raw="#2E6DB4",
        smoothed="#C97A0A",
        fit="#2CA02C",
        derivative="#1E8C5E",
        derivative_pos="#E07070",
        derivative_smooth="#136644",
        residual_above="#F0C060",
        residual_below="#70D0A0",
    ),
    "Teal": PaletteConfig(
        name="Teal",
        accent="#0D9488",   # teal-600
        raw="#00897B",
        smoothed="#26A69A",
        fit="#004D40",
        derivative="#00695C",
        derivative_pos="#EF5350",
        derivative_smooth="#00796B",
        residual_above="#FFB74D",
        residual_below="#4DB6AC",
    ),
    "Warm": PaletteConfig(
        name="Warm",
        accent="#D97706",   # amber-600
        raw="#E65100",
        smoothed="#FF8F00",
        fit="#BF360C",
        derivative="#D84315",
        derivative_pos="#C62828",
        derivative_smooth="#A1887F",
        residual_above="#FFE082",
        residual_below="#FFAB91",
    ),
    "Monochrome": PaletteConfig(
        name="Monochrome",
        accent="#374151",   # gray-700
        raw="#424242",
        smoothed="#757575",
        fit="#1565C0",
        derivative="#616161",
        derivative_pos="#EF5350",
        derivative_smooth="#9E9E9E",
        residual_above="#E0E0E0",
        residual_below="#BDBDBD",
    ),
    "Forest": PaletteConfig(
        name="Forest",
        accent="#16A34A",   # green-600
        raw="#2E7D32",
        smoothed="#558B2F",
        fit="#33691E",
        derivative="#388E3C",
        derivative_pos="#D84315",
        derivative_smooth="#1B5E20",
        residual_above="#A5D6A7",
        residual_below="#81C784",
    ),
}

File: frontend/src/lib/palettes.ts
/**
 * Chart colour palettes — the single source of truth for trace colours.
 *
 * Rendering is fully client-side, so palettes live here (they used to live in
 * the backend `viz/palettes.py`). Each palette assigns a colour to every chart
 * trace; `accent` is reserved for interactive chrome (buttons, focus rings,
 * titles) and is never used on a trace.
 */

export interface ChartPalette {
  name: string;
  /** Primary UI accent — chrome only, never a chart trace. */
  accent: string;
  raw: string;
  smoothed: string;
  fit: string;
  fitLinear: string;
  band: string;
  derivative: string;
  derivativePos: string;
  derivativeSmooth: string;
  residualAbove: string;
  residualBelow: string;
}

export const PALETTES: Record<string, ChartPalette> = {
  Classic: {
    name: "Classic",
    accent: "#2563EB",
    raw: "#2E6DB4",
    smoothed: "#C97A0A",
    fit: "#2CA02C",
    fitLinear: "#9467BD",
    band: "#94A3B8",
    derivative: "#1E8C5E",
    derivativePos: "#E07070",
    derivativeSmooth: "#136644",
    residualAbove: "#F0C060",
    residualBelow: "#70D0A0",
  },
  Teal: {
    name: "Teal",
    accent: "#0D9488",
    raw: "#00897B",
    smoothed: "#26A69A",
    fit: "#004D40",
    fitLinear: "#5E35B1",
    band: "#80CBC4",
    derivative: "#00695C",
    derivativePos: "#EF5350",
    derivativeSmooth: "#00796B",
    residualAbove: "#FFB74D",
    residualBelow: "#4DB6AC",
  },
  Warm: {
    name: "Warm",
    accent: "#D97706",
    raw: "#E65100",
    smoothed: "#FF8F00",
    fit: "#BF360C",
    fitLinear: "#6A1B9A",
    band: "#FFCC80",
    derivative: "#D84315",
    derivativePos: "#C62828",
    derivativeSmooth: "#A1887F",
    residualAbove: "#FFE082",
    residualBelow: "#FFAB91",
  },
  Monochrome: {
    name: "Monochrome",
    accent: "#374151",
    raw: "#424242",
    smoothed: "#757575",
    fit: "#1565C0",
    fitLinear: "#6D4C41",
    band: "#BDBDBD",
    derivative: "#616161",
    derivativePos: "#EF5350",
    derivativeSmooth: "#9E9E9E",
    residualAbove: "#E0E0E0",
    residualBelow: "#BDBDBD",
  },
  Forest: {
    name: "Forest",
    accent: "#16A34A",
    raw: "#2E7D32",
    smoothed: "#558B2F",
    fit: "#33691E",
    fitLinear: "#6A1B9A",
    band: "#A5D6A7",
    derivative: "#388E3C",
    derivativePos: "#D84315",
    derivativeSmooth: "#1B5E20",
    residualAbove: "#A5D6A7",
    residualBelow: "#81C784",
  },
};

export const PALETTE_NAMES: string[] = Object.keys(PALETTES);

/** Three representative swatches per palette for the picker preview. */
export const PALETTE_PREVIEWS: Record<string, string[]> = Object.fromEntries(
  Object.values(PALETTES).map((p) => [p.name, [p.raw, p.smoothed, p.fit]]),
);

export function getPalette(name: string): ChartPalette {
  return PALETTES[name] ?? PALETTES.Classic;
}

export function getPaletteAccent(name: string): string {
  return getPalette(name).accent;
}

/** Theme-neutral colours (axes, gridlines, text, tooltip) per light/dark mode. */
export interface ChartTheme {
  axis: string;
  grid: string;
  text: string;
  mutedText: string;
  reference: string;
  tooltipBg: string;
  tooltipText: string;
  tooltipBorder: string;
}

const LIGHT_THEME: ChartTheme = {
  axis: "#94A3B8",
  grid: "#E2E8F0",
  text: "#334155",
  mutedText: "#64748B",
  reference: "#94A3B8",
  tooltipBg: "#FFFFFF",
  tooltipText: "#1E293B",
  tooltipBorder: "#E2E8F0",
};

const DARK_THEME: ChartTheme = {
  axis: "#475569",
  grid: "#334155",
  text: "#CBD5E1",
  mutedText: "#94A3B8",
  reference: "#64748B",
  tooltipBg: "#1E293B",
  tooltipText: "#F1F5F9",
  tooltipBorder: "#334155",
};

export function getChartTheme(dark: boolean): ChartTheme {
  return dark ? DARK_THEME : LIGHT_THEME;
}

/** Convert a `#RRGGBB` hex colour to an `rgba(...)` string at the given alpha. */
export function hexToRgba(hex: string, alpha: number): string {
  const h = hex.replace("#", "");
  const r = parseInt(h.slice(0, 2), 16);
  const g = parseInt(h.slice(2, 4), 16);
  const b = parseInt(h.slice(4, 6), 16);
  return `rgba(${r}, ${g}, ${b}, ${alpha})`;
}
